man/boobill.1
.\" -*- coding: utf-8 -*-
.\" This file was generated automatically by tools/make_man.sh.
.TH BOOBILL 1 "13 February 2020" "boobill 2\&.0"
.SH NAME
boobill \- get/download documents and bills
.SH SYNOPSIS
.B boobill
[\-dqv] [\-b \fIbackends\fR] [\-cnfs] [\fIcommand\fR [\fIarguments\fR..]]
.br
.B boobill
[\-\-help] [\-\-version]

.SH DESCRIPTION
.LP

Console application allowing to get/download documents and bills.

.SS Supported websites:
* amazon (Amazon)
.br
* ameli (le site de l'Assurance Maladie en ligne)
.br
* amelipro (Ameli website: French Health Insurance for Professionals)
.br
* axabanque (AXA Banque)
.br
* banquepopulaire (Banque Populaire)
.br
* bnporc (BNP Paribas)
.br
* bolden (Bolden)
.br
* bouygues (Bouygues Télécom)
.br
* bp (La Banque Postale)
.br
* caissedepargne (Caisse d'Épargne)
.br
* cesu (Le Cesu est une offre simplifiée pour déclarer facilement la rémunération de votre salarié à domicile pour des activités de service à la personne.)
.br
* cityscoot (Le scooter électrique en libre\-service, sans bornes.)
.br
* creditmutuel (Crédit Mutuel)
.br
* edf (EDF)
.br
* ekwateur (ekwateur website)
.br
* ensap (ensap website)
.br
* freemobile (Free Mobile website)
.br
* infomaniak (Infomaniak)
.br
* ing (ING France)
.br
* lampiris (French electricity provider Lampiris.fr)
.br
* lcl (LCL)
.br
* ldlc (ldlc website)
.br
* lucca (Lucca RH)
.br
* materielnet (Materiel.net)
.br
* myfoncia (Foncia billing capabilities)
.br
* onlinenet (Online.net)
.br
* orange (Orange French mobile phone provider)
.br
* ovh (Ovh)
.br
* pajemploi (Pajemploi est une offre de service du réseau des Urssaf, destinée à simplifier les formalités administratives pour les parents employeurs qui font garder leur(s) enfant(s) par une assistante maternelle agréée ou une garde d’enfants à domicile.)
.br
* poivy (Poivy website)
.br
* societegenerale (Société Générale)
.br
* trainline (trainline)
.SH BOOBILL COMMANDS
.TP
\fBbalance\fR [\fIID\fR]
.br
Get balance of subscriptions.
.br
If no ID given, display balance of all backends.
.TP
\fBbills\fR [\fIID\fR]
.br
Get the list of bills documents for subscriptions.
.br
If no ID given, display bills of all backends
.br

.br
Default is limited to 10 results.
.TP
\fBdebug\fR
.br
Launch a debug Python shell
.TP
\fBdetails\fR [\fIID\fR]
.br
Get details of subscriptions.
.br
If no ID given, display all details of all backends.
.TP
\fBdocuments\fR [\fIID\fR]
.br
Get the list of documents for subscriptions.
.br
If no ID given, display documents of all backends
.br

.br
Default is limited to 10 results.
.TP
\fBdownload\fR [\fIDOC_ID\fR | all] [\fIFILENAME\fR]
.br
download DOC_ID [FILENAME]
.br

.br
download the document
.br
DOC_ID is the identifier of the document (hint: try documents command)
.br
FILENAME is where to write the file. If FILENAME is '\-',
.br
the file is written to stdout.
.br

.br
download all [SUB_ID]
.br

.br
You can use special word "all" and download all documents of
.br
subscription identified by SUB_ID.
.br
If SUB_ID is not given, download documents of all subscriptions.
.TP
\fBdownload_pdf\fR [id | all]
.br
download function with forced PDF conversion.
.TP
\fBhistory\fR [\fIID\fR]
.br
Get the history of subscriptions.
.br
If no ID given, display histories of all backends.
.br

.br
Default is limited to 10 results.
.TP
\fBprofile\fR
.br
Display detailed information about person or company.
.TP
\fBsubscriptions\fR
.br
List all subscriptions.
.SH WEBOOB COMMANDS
.TP
\fBbackends\fR [\fIACTION\fR] [\fIBACKEND_NAME\fR]...
.br
Select used backends.
.br

.br
ACTION is one of the following (default: list):
.br
* enable         enable given backends
.br
* disable        disable given backends
.br
* only           enable given backends and disable the others
.br
* list           list backends
.br
* add            add a backend
.br
* register       register a new account on a website
.br
* edit           edit a backend
.br
* remove         remove a backend
.br
* list\-modules   list modules
.TP
\fBcd\fR [\fIPATH\fR]
.br
Follow a path.
.br
".." is a special case and goes up one directory.
.br
"" is a special case and goes home.
.TP
\fBcondition\fR [\fIEXPRESSION\fR | off]
.br
If an argument is given, set the condition expression used to filter the results. See CONDITION section for more details and the expression.
.br
If the "off" value is given, conditional filtering is disabled.
.br

.br
If no argument is given, print the current condition expression.
.TP
\fBcount\fR [\fINUMBER\fR | off]
.br
If an argument is given, set the maximum number of results fetched.
.br
NUMBER must be at least 1.
.br
"off" value disables counting, and allows infinite searches.
.br

.br
If no argument is given, print the current count value.
.TP
\fBformatter\fR [list | \fIFORMATTER\fR [\fICOMMAND\fR] | option \fIOPTION_NAME\fR [on | off]]
.br
If a FORMATTER is given, set the formatter to use.
.br
You can add a COMMAND to apply the formatter change only to
.br
a given command.
.br

.br
If the argument is "list", print the available formatters.
.br

.br
If the argument is "option", set the formatter options.
.br
Valid options are: header, keys.
.br
If on/off value is given, set the value of the option.
.br
If not, print the current value for the option.
.br

.br
If no argument is given, print the current formatter.
.TP
\fBlogging\fR [\fILEVEL\fR]
.br
Set logging level.
.br

.br
Availables: debug, info, warning, error.
.br
* quiet is an alias for error
.br
* default is an alias for warning
.TP
\fBls\fR [\-d] [\-\fIU\fR] [\fIPATH\fR]
.br
List objects in current path.
.br
If an argument is given, list the specified path.
.br
Use \-U option to not sort results. It allows you to use a "fast path" to
.br
return results as soon as possible.
.br
Use \-d option to display information about a collection (and to not
.br
display the content of it). It has the same behavior than the well
.br
known UNIX "ls" command.
.br

.br
Default is limited to 40 results.
.TP
\fBquit\fR
.br
Quit the application.
.TP
\fBselect\fR [\fIFIELD_NAME\fR]... | "$direct" | "$full"
.br
If an argument is given, set the selected fields.
.br
$direct selects all fields loaded in one http request.
.br
$full selects all fields using as much http requests as necessary.
.br

.br
If no argument is given, print the currently selected fields.

.SH OPTIONS
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-b BACKENDS\fR, \fB\-\-backends=BACKENDS\fR
what backend(s) to enable (comma separated)
.TP
\fB\-e EXCLUDE_BACKENDS\fR, \fB\-\-exclude\-backends=EXCLUDE_BACKENDS\fR
what backend(s) to exclude (comma separated)
.TP
\fB\-I\fR, \fB\-\-insecure\fR
do not validate SSL
.TP
\fB\-\-nss\fR
Use NSS instead of OpenSSL
.TP
\fB\-\-auto\-update\fR
Automatically check for updates when a bug in a module is encountered

.SH LOGGING OPTIONS
.TP
\fB\-d\fR, \fB\-\-debug\fR
display debug messages. Set up it twice to more verbosity
.TP
\fB\-q\fR, \fB\-\-quiet\fR
display only error messages
.TP
\fB\-v\fR, \fB\-\-verbose\fR
display info messages
.TP
\fB\-\-logging\-file=LOGGING_FILE\fR
file to save logs
.TP
\fB\-a\fR, \fB\-\-save\-responses\fR
save every response
.TP
\fB\-\-export\-session\fR
log browser session cookies after login

.SH RESULTS OPTIONS
.TP
\fB\-c CONDITION\fR, \fB\-\-condition=CONDITION\fR
filter result items to display given a boolean expression. See CONDITION section
for the syntax
.TP
\fB\-n COUNT\fR, \fB\-\-count=COUNT\fR
limit number of results (from each backends)
.TP
\fB\-s SELECT\fR, \fB\-\-select=SELECT\fR
select result item keys to display (comma separated)

.SH FORMATTING OPTIONS
.TP
\fB\-f FORMATTER\fR, \fB\-\-formatter=FORMATTER\fR
select output formatter (csv, htmltable, json, json_line, multiline, simple,
subscriptions, table, webkit)
.TP
\fB\-\-no\-header\fR
do not display header
.TP
\fB\-\-no\-keys\fR
do not display item keys
.TP
\fB\-O OUTFILE\fR, \fB\-\-outfile=OUTFILE\fR
file to export result

.SH CONDITION
The \-c and \-\-condition is a flexible way to filter and get only interesting results. It supports conditions on numerical values, dates, durations and strings. Dates are given in YYYY\-MM\-DD or YYYY\-MM\-DD HH:MM format. Durations look like XhYmZs where X, Y and Z are integers. Any of them may be omitted. For instance, YmZs, XhZs or Ym are accepted.
The syntax of one expression is "\fBfield operator value\fR". The field to test is always the left member of the expression.
.LP
The field is a member of the objects returned by the command. For example, a bank account has "balance", "coming" or "label" fields.
.SS The following operators are supported:
.TP
=
Test if object.field is equal to the value.
.TP
!=
Test if object.field is not equal to the value.
.TP
>
Test if object.field is greater than the value. If object.field is date, return true if value is before that object.field.
.TP
<
Test if object.field is less than the value. If object.field is date, return true if value is after that object.field.
.TP
|
This operator is available only for string fields. It works like the Unix standard \fBgrep\fR command, and returns True if the pattern specified in the value is in object.field.
.SS Expression combination
.LP
You can make a expression combinations with the keywords \fB" AND "\fR, \fB" OR "\fR an \fB" LIMIT "\fR.
.LP
The \fBLIMIT\fR keyword can be used to limit the number of items upon which running the expression. \fBLIMIT\fR can only be placed at the end of the expression followed by the number of elements you want.
.SS Examples:
.nf
.B boobank ls \-\-condition 'label=Livret A'
.fi
Display only the "Livret A" account.
.PP
.nf
.B boobank ls \-\-condition 'balance>10000'
.fi
Display accounts with a lot of money.
.PP
.nf
.B boobank history account@backend \-\-condition 'label|rewe'
.fi
Get transactions containing "rewe".
.PP
.nf
.B boobank history account@backend \-\-condition 'date>2013\-12\-01 AND date<2013\-12\-09'
.fi
Get transactions betweens the 2th December and 8th December 2013.
.PP
.nf
.B boobank history account@backend \-\-condition 'date>2013\-12\-01  LIMIT 10'
.fi
Get transactions after the 2th December in the last 10 transactions

.SH COPYRIGHT
Copyright(C) 2012-2020 Florent Fourcot
.LP
For full copyright information see the COPYING file in the weboob package.
.LP
.RE
.SH FILES
"~/.config/weboob/backends" 

.SH SEE ALSO
Home page: http://weboob.org/applications/boobill


ChangeLog
Weboob 2.0 (2020-02-13)

	Applications: boobank
	* keep Value for TransferStep and AddRecipientStep
	* declare utf-8 in OFX formatter
	* give account object to iter_transfer_recipient

	Applications: boobill
	* using a persistent storage for browser state

	Applications: boobsize
	* fix name

	Applications: translaboob
	* use babel to detect language names and pass ISO-639 codes

	Applications: weboobdebug
	* weboob-debug: fix name

	Applications: weboobmain
	* move scripts/weboob to a dedicated dir in weboob/applications

	Applications: weboobrepos
	* weboob-repos: port to python3

	Browser
	* set a parent logger to AbstractElement
	* Allows SeleniumBrowser to use MANUAL proxies
	* AUTHENTICATION_METHODS set in init
	* delete prefix 'handle_' on authentication methods
	* clean authentication config keys after login
	* make variable name more understandable
	* make get_expire stronger
	* TWOFA_DURATION for the state expire date
	* better names
	* set expire from logged_date
	* make authentication methods more flexible
	* handle both 2FA and regular login methods
	* add possibility to clear specific cookies
	* handle NeedInteractiveFor2FA
	* add documentation about otp, sms and resume
	* add TwoFactorBrowser to handle multiple 2FA
	* Set default logging level to ERROR for selenium
	* Added the ability to overrride *Elements' env copy
	* Allow passing custom headers in Form.submit()
	* set attribute 'weboob' on browser
	* Normalize encodings
	* BrowserTooManyRequests: Remove class attributes for date helpers
	* BrowserTooManyRequests: Raise a TypeError if the next_try is not a datetime
	* BrowserTooManyRequests: Check whether the next_try is None after the date check
	* BrowserTooManyRequests: Add message parameter
	* [weboob/browser/browser+pages] Fixed doctest BASEURL
	* add some helpers to TooManyRequests
	* add rate limiting exception
	* add build_refresh_token_parameters method
	* python3 compatibility
	* send redirect_uri when using the refresh_token
	* allow multiple inheritance with AbstractBrowser
	* Implemented wait_xpath_invisible
	* Avoid crash when no BASEURL is set
	* add klass attribute to decorated @method
	* give state to the module at initialization
	* introduce BrowserParamURL for passing params implictly
	* Implement the empty_xpath attribute
	* accept new parameter proxy_headers (but don't use it)
	* Browser takes a new parameter 'proxy_headers'
	* if use of refresh_token fails, forget it
	* support when token_type key is missing
	* if access_token_expire is None, the access_token has no expiration
	* in case we got a 401 response, forget the access_token
	* fix compatibility with python3
	* Handle generic error for bad url callback
	* Repair refresh token
	* StatesMixin cookies should be str
	* OAuth2PKCEMixin codes shouldn't be bytes
	* Rewrite/improve PagesBrowser example

	Browser: Filters
	* Add the format .Italian() to CleanDecimal
	* New keyword argument to remove accented characters on CleanText filter.
	* Changed Capitalize filter to Title
	* Updated & Reordered __all__
	* Add dates to FormValue
	* Add color to FormValue
	* Add number/range type to FormValue, and tests
	* MultiJoin accepts types other than str
	* Regexp can use regex library
	* Allow all dateutil.parser arguments in DateTime
	* Strict dates by default
	* Catch incomplete dates
	* Explicit error message
	* Add MultiJoin
	* Join can use default argument
	* Add doctests to Join
	* TableCell(colspan=True) is now the only behavior desired
	* add FromTimestamp to convert to datetime
	* Fix error message for FormValue
	* Add new types for FormValue
	* Add MapIn filter, that allow the pattern of a selected value to another value using a dict.
	* Reorganize imports / add missing to __all__
	* Add Coalesce

	Capabilities
	* rename accepted_doc_types attribute of CapDocument
	* Add KIID to DocumentTypes
	* Added new capability CapBands
	* Keep only PER
	* add PER account type
	* HousingPhoto inherits BaseImage
	* use PostalAddress and GeoCoords types where possible
	* introduce weboob.capabilities.address for a few object types
	* Add firstname and lastname field
	* add capability_to_string function to get capability name
	* Add new SAR currency
	* fix missing TransferInvalidOTP export
	* add TransferInvalidOTP exception
	* use StringField standard default value for transfer beneficiary label and number
	* Indentation cosmetic changes
	* Add new Investment attributes
	* Add new document type 'income_tax'
	* Add field 'original_gross_amount' to Transaction
	* Add field 'gross_amount' to Transaction
	* Remove performance_history default value
	* Added new attributes for employee savings accounts
	* Add RecaptchaV3Job to captcha capability
	* add support for commission in currency in bank/Transaction
	* add new add recipient exception for invalid otp
	* add has_file field to Document class
	* Add the relationship between the credentials owner (PSU) and the account
	* Add bdate. Represent the date that the transaction appear in the website.
	* prevent casting id from None to 'None' in constructor

	Documentation
	* add pip install
	* fix compatibility with sphinx 2.0
	* do not use deprecated defindex.html anymore
	* Use correct word

	General
	* New airparif module (CapGauge)
	* New aviva module (CapBank, CapBankWealth)
	* New cesu module (CapDocument)
	* New degiro module (CapBank, CapBankWealth)
	* New hsbchk module (CapBank)
	* New lendosphere module (CapBank, CapBankWealth)
	* New lucca module (CapCalendarEvent, CapDocument)
	* New lunchr module (CapBank)
	* New metalarchives module (CapBands)
	* New pajemploi module (CapDocument)
	* New ticketscesu module (CapBank)
	* New wiseed module (CapBank, CapBankWealth)
	* Deleted biplan module
	* Deleted figgo module
	* Deleted lolix module
	* Deleted wellsfargo module

	General: Core
	* send request_information to modules in all case, except if stdout is not a tty
	* handle error NeedInteractiveFor2FA

weboob/applications/boobill/boobill.py
# -*- coding: utf-8 -*-

# Copyright(C) 2012-2013 Florent Fourcot
#
# This file is part of weboob.
#
# weboob is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# weboob is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with weboob. If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

from decimal import Decimal
import sys

from weboob.capabilities.bill import CapDocument, Detail, Subscription
from weboob.capabilities.profile import CapProfile
from weboob.tools.application.repl import ReplApplication, defaultcount
from weboob.tools.application.formatters.iformatter import PrettyFormatter
from weboob.tools.application.base import MoreResultsAvailable
from weboob.tools.application.captcha import CaptchaMixin
from weboob.core import CallErrors
from weboob.exceptions import CaptchaQuestion
from weboob.capabilities.captcha import exception_to_job


__all__ = ['Boobill']


class SubscriptionsFormatter(PrettyFormatter):
    MANDATORY_FIELDS = ('id', 'label')

    def get_title(self, obj):
        if obj.renewdate:
            return u"%s - %s" % (obj.label, obj.renewdate.strftime('%d/%m/%y'))
        return obj.label


class Boobill(CaptchaMixin, ReplApplication):
    APPNAME = 'boobill'
    VERSION = '2.1'
    COPYRIGHT = 'Copyright(C) 2012-YEAR Florent Fourcot'
    DESCRIPTION = 'Console application allowing to get/download documents and bills.'
    SHORT_DESCRIPTION = "get/download documents and bills"
    CAPS = CapDocument
    COLLECTION_OBJECTS = (Subscription, )
    EXTRA_FORMATTERS = {'subscriptions':   SubscriptionsFormatter,
                        }
    DEFAULT_FORMATTER = 'table'
    COMMANDS_FORMATTERS = {'subscriptions':   'subscriptions',
                           'ls':              'subscriptions',
                          }

    def load_default_backends(self):
        self.load_backends(CapDocument, storage=self.create_storage())

    def main(self, argv):
        self.load_config()
        return ReplApplication.main(self, argv)

    def exec_method(self, id, method):
        l = []
        id, backend_name = self.parse_id(id)

        if not id:
            for subscrib in self.get_object_list('iter_subscription'):
                l.append((subscrib.id, subscrib.backend))
        else:
            l.append((id, backend_name))

        more_results = []
        not_implemented = []
        self.start_format()
        for id, backend in l:
            names = (backend,) if backend is not None else None
            try:
                for result in self.do(method, id, backends=names):
                    self.format(result)
            except CallErrors as errors:
                for backend, error, backtrace in errors:
                    if isinstance(error, MoreResultsAvailable):
                        more_results.append(id + u'@' + backend.name)
                    elif isinstance(error, NotImplementedError):
                        if backend not in not_implemented:
                            not_implemented.append(backend)
                    else:
                        self.bcall_error_handler(backend, error, backtrace)

        if len(more_results) > 0:
            print('Hint: There are more results available for %s (use option -n or count command)' % (', '.join(more_results)), file=self.stderr)
        for backend in not_implemented:
            print(u'Error(%s): This feature is not supported yet by this backend.' % backend.name, file=self.stderr)

    def do_subscriptions(self, line):
        """
        subscriptions

        List all subscriptions.
        """
        return self.do_ls(line)

    def do_details(self, id):
        """
        details [ID]

        Get details of subscriptions.
        If no ID given, display all details of all backends.
        """
        l = []
        id, backend_name = self.parse_id(id)

        if not id:
            for subscrib in self.get_object_list('iter_subscription'):
                l.append((subscrib.id, subscrib.backend))
        else:
            l.append((id, backend_name))

        for id, backend in l:
            names = (backend,) if backend is not None else None
            # XXX: should be generated by backend? -Flo
            # XXX: no, but you should do it in a specific formatter -romain
            # TODO: do it, and use exec_method here. Code is obsolete
            mysum = Detail()
            mysum.label = u"Sum"
            mysum.infos = u"Generated by boobill"
            mysum.price = Decimal("0.")

            self.start_format()
            for detail in self.do('get_details', id, backends=names):
                self.format(detail)
                mysum.price = detail.price + mysum.price

            self.format(mysum)

    def do_balance(self, id):
        """
        balance [ID]

        Get balance of subscriptions.
        If no ID given, display balance of all backends.
        """

        self.exec_method(id, 'get_balance')

    @defaultcount(10)
    def do_history(self, id):
        """
        history [ID]

        Get the history of subscriptions.
        If no ID given, display histories of all backends.
        """
        self.exec_method(id, 'iter_bills_history')

    @defaultcount(10)
    def do_documents(self, id):
        """
        documents [ID]

        Get the list of documents for subscriptions.
        If no ID given, display documents of all backends
        """
        self.exec_method(id, 'iter_documents')

    @defaultcount(10)
    def do_bills(self, id):
        """
        bills [ID]

        Get the list of bills documents for subscriptions.
        If no ID given, display bills of all backends
        """
        self.exec_method(id, 'iter_bills')

    def do_download(self, line, force_pdf=False):
        """
        download [DOC_ID | all] [FILENAME]

        download DOC_ID [FILENAME]

        download the document
        DOC_ID is the identifier of the document (hint: try documents command)
        FILENAME is where to write the file. If FILENAME is '-',
        the file is written to stdout.

        download all [SUB_ID]

        You can use special word "all" and download all documents of
        subscription identified by SUB_ID.
        If SUB_ID is not given, download documents of all subscriptions.
        """
        id, dest = self.parse_command_args(line, 2, 1)
        id, backend_name = self.parse_id(id)
        if not id:
            print('Error: please give a document ID (hint: use documents command)', file=self.stderr)
            return 2

        if id == 'all':
            return self.download_all(dest, force_pdf)

        names = (backend_name,) if backend_name is not None else None

        document, = self.do('get_document', id, backends=names)
        if not document:
            print('Error: document not found')
            return 1

        if dest is None:
            dest = id + "." + (document.format if not force_pdf else 'pdf')

        for buf in self.do('download_document' if not force_pdf else 'download_document_pdf', document, backends=names):
            if buf:
                if dest == "-":
                    if sys.version_info.major >= 3:
                        self.stdout.buffer.write(buf)
                    else:
                        self.stdout.stream.write(buf)
                else:
                    try:
                        with open(dest, 'wb') as f:
                            f.write(buf)
                    except IOError as e:
                        print('Unable to write document in "%s": %s' % (dest, e), file=self.stderr)
                        return 1
                return

    def do_download_pdf(self, line):
        """
        download_pdf [id | all]

        download function with forced PDF conversion.
        """

        return self.do_download(line, force_pdf=True)

    def download_all(self, sub_id, force_pdf):
        if sub_id:
            sub_id, backend_name = self.parse_id(sub_id)
            names = (backend_name,) if backend_name else None
            subscription, = self.do('get_subscription', sub_id, backends=names)
            if not self.download_subscription(subscription, force_pdf):
                return 1
        else:
            for subscription in self.do('iter_subscription'):
                if not self.download_subscription(subscription, force_pdf):
                    return 1

    def download_subscription(self, subscription, force_pdf):
        for document in self.do('iter_documents', subscription, backends=(subscription.backend,)):
            if not self.download_doc(document, force_pdf):
                return False
        return True

    def download_doc(self, document, force_pdf):
        if force_pdf:
            method = 'download_document_pdf'
        else:
            method = 'download_document'

        dest = document.id + "." + (document.format if not force_pdf else 'pdf')

        for buf in self.do(method, document, backends=(document.backend,)):
            if buf:
                try:
                    with open(dest, 'wb') as f:
                        f.write(buf)
                except IOError as e:
                    print('Unable to write bill in "%s": %s' % (dest, e), file=self.stderr)
                    return False
        return True

    def do_profile(self, line):
        """
        profile

        Display detailed information about person or company.
        """
        self.start_format()
        for profile in self.do('get_profile', caps=CapProfile):
            self.format(profile)

    def bcall_error_handler(self, backend, error, backtrace):
        """
        Handler for an exception inside the CallErrors exception.

        This method can be overridden to support more exceptions types.
        """
        if isinstance(error, CaptchaQuestion):
            if not self.captcha_weboob.count_backends():
                print('Error(%s): Site requires solving a CAPTCHA but no CapCaptchaSolver backends were configured' % backend.name,
                      file=self.stderr)
                return False

            print('Info(%s): Encountered CAPTCHA, please wait for its resolution, it can take dozens of seconds' % backend.name, file=self.stderr)
            job = exception_to_job(error)
            self.solve_captcha(job, backend)
            return False

        return super(ReplApplication, self).bcall_error_handler(backend, error, backtrace)
